File: config/initializers/sentry.rb
# Sentry error tracking configuration
# https://docs.sentry.io/platforms/ruby/guides/rails/
# https://docs.sentry.io/platforms/ruby/guides/rails/configuration/options/

Sentry.init do |config|
  # ====================
  # Core Options
  # ====================

  # Set the Sentry DSN from environment variables
  # If not provided, SDK will try to read from SENTRY_DSN env var
  config.dsn = ENV["SENTRY_DSN"]

  # Enable Sentry only in production and staging environments
  config.enabled_environments = %w[production staging]

  # Set the current environment
  config.environment = Rails.env

  # Set release version for tracking which version has bugs
  # Sentry automatically detects releases from git, Heroku, Capistrano, etc.
  config.release = ENV["APP_VERSION"] || "framefox-connect@#{ENV['HEROKU_SLUG_COMMIT']&.slice(0, 7)}"

  # ====================
  # Sampling Options
  # ====================

  # Error event sample rate (0.0 to 1.0)
  # 1.0 means 100% of error events will be sent
  config.sample_rate = 1.0

  # Performance monitoring: transaction sample rate (0.0 to 1.0)
  # 0.1 means 10% of transactions will be sent to Sentry
  config.traces_sample_rate = ENV.fetch("SENTRY_TRACES_SAMPLE_RATE", 0.1).to_f

  # ====================
  # Data & Privacy Options
  # ====================

  # Send personally identifiable information (PII) with events
  # WARNING: When enabled, request data, user IPs, and other PII will be sent to Sentry
  # Sentry will automatically respect Rails.application.config.filter_parameters
  config.send_default_pii = true

  # Send diagnostic client reports about dropped events (recommended)
  config.send_client_reports = true

  # Send gem/module dependency information with events
  config.send_modules = true

  # ====================
  # Breadcrumbs
  # ====================

  # Maximum number of breadcrumbs to store (default: 100)
  config.max_breadcrumbs = 50

  # Breadcrumbs are a trail of events that happened prior to an error
  # :active_support_logger - Rails-specific info via ActiveSupport instrumentation
  # :http_logger - Captures requests made with net/http library
  config.breadcrumbs_logger = [ :active_support_logger, :http_logger ]

  # ====================
  # Exception Filtering
  # ====================

  # Exclude certain exception types that are not actionable errors
  config.excluded_exceptions += [
    "ActionController::RoutingError",
    "ActiveRecord::RecordNotFound",
    "ActionController::InvalidAuthenticityToken",
    "ActionController::UnknownFormat",
    "ActionDispatch::Http::MimeNegotiation::InvalidType"
  ]

  # ====================
  # Callbacks & Hooks
  # ====================

  # Clean up backtrace before sending to Sentry (removes gem/framework noise)
  config.backtrace_cleanup_callback = lambda do |backtrace|
    Rails.backtrace_cleaner.clean(backtrace)
  end

  # Modify or filter events before sending to Sentry
  config.before_send = lambda do |event, hint|
    # Add custom user context if logged in
    if defined?(Current) && Current.user
      event.user = {
        id: Current.user.id,
        email: Current.user.email,
        username: Current.user.email
      }
    end

    # Add custom tags
    event.tags[:subdomain] = Current.subdomain if defined?(Current) && Current.subdomain

    # Return event to send it, or return nil to drop it
    event
  end

  # ====================
  # Rails Integration
  # ====================

  # Automatically track user sessions in request/response cycles
  config.auto_session_tracking = true

  # Note: Performance tracing subscribers are automatically enabled in sentry-rails 6.x
  # The config.rails.tracing_subscribers option is only available in Sentry 7.0+
  # Sentry 6.x automatically instruments:
  # - Active Record queries
  # - Action Controller requests
  # - Action View rendering
  # - Active Storage operations

  # ====================
  # Background Jobs (Sidekiq)
  # ====================

  # Only report Sidekiq jobs after all retries have been exhausted
  config.sidekiq.report_after_job_retries = true

  # ====================
  # Debug & Logging
  # ====================

  # Enable debug mode to see SDK errors with backtrace (not recommended for production)
  # config.debug = true

  # Adjust SDK logger level if you need more output (not recommended for production)
  # config.logger.level = ::Logger::DEBUG
end

File: app/javascript/utils/sentry.js
// Sentry frontend error tracking
// https://docs.sentry.io/platforms/javascript/

import * as Sentry from "@sentry/browser";

// Initialize Sentry only if DSN is configured
const sentryDsn = document.querySelector('meta[name="sentry-dsn"]')?.content;

if (sentryDsn) {
  Sentry.init({
    dsn: sentryDsn,
    
    // Set the environment (production, staging, development)
    environment: document.querySelector('meta[name="rails-env"]')?.content || 'development',
    
    // Set release version to track which version has bugs
    release: document.querySelector('meta[name="app-version"]')?.content,
    
    // Tracing sample rate (0.0 to 1.0)
    // This determines the percentage of transactions sent to Sentry
    tracesSampleRate: parseFloat(document.querySelector('meta[name="sentry-traces-sample-rate"]')?.content || '0.1'),
    
    // Capture unhandled promise rejections
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.replayIntegration({
        // Capture 10% of all sessions for replay
        sessionSampleRate: 0.1,
        // Capture 100% of sessions with errors for replay
        errorSampleRate: 1.0,
      }),
    ],
    
    // Don't send errors from localhost/development
    beforeSend(event, hint) {
      // Don't send events in development unless explicitly enabled
      if (window.location.hostname === 'localhost' || window.location.hostname === '127.0.0.1') {
        return null;
      }
      return event;
    },
    
    // Filter out certain errors that aren't actionable
    ignoreErrors: [
      // Browser extension errors
      'Non-Error promise rejection captured',
      'ResizeObserver loop limit exceeded',
      // Network errors
      'NetworkError',
      'Network request failed',
      // Script loading errors from browser extensions
      /^Loading chunk \d+ failed/,
    ],
  });

  // Set user context if available
  const userEmail = document.querySelector('meta[name="current-user-email"]')?.content;
  const userId = document.querySelector('meta[name="current-user-id"]')?.content;
  
  if (userEmail && userId) {
    Sentry.setUser({
      id: userId,
      email: userEmail,
    });
  }

  // Add custom context
  const subdomain = document.querySelector('meta[name="subdomain"]')?.content;
  if (subdomain) {
    Sentry.setTag('subdomain', subdomain);
  }
  
  console.log('Sentry initialized for frontend error tracking');
}

// Export Sentry for manual error reporting if needed
export default Sentry;

